dvc/project.py

```diff
@@ -1041,47 +1041,74 @@ def _read_metric(self, path, typ=None, xpath=None):
 
         return ret
 
-    def _find_output_by_path(self, path, outs=None):
+    def _find_output_by_path(self, path, outs=None, recursive=False):
         from dvc.exceptions import OutputDuplicationError
 
         if not outs:
             astages = self.active_stages()
             outs = [out for stage in astages for out in stage.outs]
 
         abs_path = os.path.abspath(path)
-        matched = [out for out in outs if out.path == abs_path]
-        stages = [out.stage.relpath for out in matched]
-        if len(stages) > 1:
-            raise OutputDuplicationError(path, stages)
+        if os.path.isdir(abs_path) and recursive:
+            matched = [out for out in outs
+                       if os.path.abspath(out.path).startswith(abs_path)]
+            stages = [out.stage.relpath for out in matched
+                      if out.path == abs_path]
+            if len(stages) > 1:
+                raise OutputDuplicationError(path, stages)
+        else:
+            matched = [out for out in outs if out.path == abs_path]
+            stages = [out.stage.relpath for out in matched]
+            if len(stages) > 1:
+                raise OutputDuplicationError(path, stages)
 
-        return matched[0] if matched else None
+        return matched if matched else None
```

Comment:
I would return an empty list instead of None here. It can spare some if's upstream.